cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java

```diff
@@ -139,27 +148,54 @@ public void startPreview(PreviewRequest<?> previewRequest) throws Exception {
       programId, previewConfig == null ? Collections.<String, String>emptyMap() : previewConfig.getRuntimeArgs(),
       false);
 
+    // Only have timer if there is a timeout setting.
+    if (previewConfig != null && previewConfig.getTimeout() != null) {
+      timer = new Timer();
+      final int runningTime =  previewConfig.getTimeout();
+      timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          try {
+            killedByTimer = true;
+            LOG.info("Stopping the preview since it has reached running time: {} mins.", runningTime);
+            stopPreview();
```

Comment:
set `killedByTimer` after `stopPreview()`, so no need to set it to false again in the exception.